ddsrouter_event/include/ddsrouter_event/impl/EventHandler.ipp

```diff
@@ -80,28 +84,38 @@ void EventHandler<Args...>::unset_callback() noexcept
 {
     std::lock_guard<std::recursive_mutex> lock(event_mutex_);
 
+    bool was_callback_set_before = false;
+
     if (is_callback_set_)
     {
+        was_callback_set_before = true;
+
+        std::lock_guard<std::recursive_mutex> lock(internal_callback_mutex_);
+
         {
             // Unsetting callback
             // Wait mutex must be taken because this variable is used in wait_for_event() wait
             std::unique_lock<std::mutex> lock(wait_mutex_);
             is_callback_set_.store(false);
         }
 
-        // Awaking every thread waiting in wait_for_event()
-        awake_all_waiting_threads_nts_();
-
         // Setting new callback (even when it will not be called)
         callback_ = DEFAULT_CALLBACK_;
-
-        // Call child methods in case they should do something when handler is disabled
-        callback_unset_nts_();
     }
     else
     {
         logWarning(DDSROUTER_HANDLER, "Unsetting callback from an EventHandler that had no callback set.")
     }
+
+    // Should only call unset if the callback was set
+    if (was_callback_set_before)
+    {
+        // Call child methods in case they should do something when handler is disabled
+        callback_unset_nts_();
+    }
+
+    // Awaking every thread waiting in wait_for_event()
+    awake_all_waiting_threads_nts_();
```

Comment:
That's right, I meant that it is not useful to notify in the case no callback was previously set (and therefore no callback is unset) because no thread should be waiting anyway.